builtin/plugins/dposv3/dpos.go

```diff
@@ -1349,15 +1355,53 @@ func ShiftDowntimeWindow(ctx contract.Context, currentHeight int64, candidates [
 	if state.Params.DowntimePeriod != 0 && (uint64(currentHeight)%state.Params.DowntimePeriod) == 0 {
 		ctx.Logger().Info("DPOS ShiftDowntimeWindow", "block", currentHeight)
 
+		downtimeSlashingEnabled := ctx.FeatureEnabled(loomchain.DPOSVersion3_4, false)
+
+		maxDowntimePercentage := defaultMaxDowntimePercentage
+		if state.Params.MaxDowntimePercentage != nil {
+			maxDowntimePercentage = state.Params.MaxDowntimePercentage.Value
+		}
+
+		maximumMissedBlocksBig := CalculateFraction(maxDowntimePercentage, loom.BigUInt{big.NewInt(int64(state.Params.DowntimePeriod))})
+		maximumMissedBlocks := maximumMissedBlocksBig.Uint64()
+
 		for _, candidate := range candidates {
-			statistic, err := GetStatistic(ctx, loom.UnmarshalAddressPB(candidate.Address))
+			candidateAddress := loom.UnmarshalAddressPB(candidate.Address)
+			statistic, err := GetStatistic(ctx, candidateAddress)
 			if err != nil {
 				if err == contract.ErrNotFound {
 					continue
 				}
 				return err
 			}
 
+			if downtimeSlashingEnabled {
+				slash := true;
+				downtime := getDowntimeRecord(ctx, statistic)
+				for i := uint64(0); i < 4; i++ {
+					if (maximumMissedBlocks >= downtime.Periods[i]) {
+						slash = false;
+						break;
+					}
+				}
+
+				if slash {
+					if err := SlashInactivity(ctx, candidateAddress); err != nil {
+						return err
+					}
+
+					// load updated statistic that was modified by the call to
+					// SlashInactivity
+					statistic, err = GetStatistic(ctx, candidateAddress)
```

Comment:
To avoid this statistic-reloading, I might make `SlashInactivity` simply return a modified statistic.

Reply:
Or you can pass the statistic into `SlashInactivity` and let it modify it.

Reply:
yeah, that's what I meant.